Add current state to workflow states table and format duration time

GilbertCherrie opened the pull request:
Fixes: https://github.com/ManageIQ/manageiq-ui-classic/issues/9189

Adds the current state to the workflow states table. Also, formats the duration time to be displays as a string in days hours minutes seconds format instead of just seconds.

Before:
<img width="1407" alt="Screenshot 2024-05-17 at 9 43 27 AM" src="https://github.com/ManageIQ/manageiq-ui-classic/assets/32444791/4704c8c0-df4c-415c-a3e2-552e55896a62">

After:
<img width="1409" alt="Screenshot 2024-05-23 at 4 53 58 PM" src="https://github.com/ManageIQ/manageiq-ui-classic/assets/32444791/1ee35fa9-16bf-4862-bd8d-86c2c1f549ec">

jeffibm commented on app/javascript/components/request-workflow-status/data.js:
```diff
@@ -59,6 +59,15 @@ export const workflowStatusData = (response) => {
     return undefined;
   }
   const rows = response.context ? rowData(response.context) : [];
+  if (response.context && response.context.State) {
+    rows.push({
+      id: `${response.context.State.Guid.toString()}+${Math.random()}`,
+      name: response.context.State.Name,
+      enteredTime: convertDate(response.context.State.EnteredTime.toString()),
+      finishedTime: convertDate(response.context.State.FinishedTime.toString()),
+      duration: response.context.State.Duration.toFixed(3).toString(),
+    });
+  }
```
Maybe, we can simplify it like this - 

```
if (response.context && response.context.State) {
  const { State } = response.context;
  rows.push({
    id: `${State.Guid}+${Math.random()}`,
    name: State.Name,
    enteredTime: convertDate(State.EnteredTime.toString()),
    finishedTime: convertDate(State.FinishedTime.toString()),
    duration: State.Duration.toFixed(3).toString(),
  });
}
```

jeffibm reviewed (commented):
Also, if possible, could you please post a before/after screenshot of the page from where this change is coming in from?

Fryguy commented:
I'm concerned this will appear like the step is done.  Wonder if we can have a leading icon that represent the current state of the state

agrare commented on app/javascript/components/request-workflow-status/data.js:
```diff
@@ -59,6 +59,15 @@ export const workflowStatusData = (response) => {
+      id: `${response.context.State.Guid.toString()}+${Math.random()}`,
+      name: response.context.State.Name,
+      enteredTime: convertDate(response.context.State.EnteredTime.toString()),
+      finishedTime: convertDate(response.context.State.FinishedTime.toString()),
```
If it is currently running it shouldn't have a FinishedTime

Fryguy replied:
nor a duration

Fryguy replied:
actually maybe it should have a duration on presentation?  not sure.

agrare replied:
Yeah I was thinking duration could be Time.now - StartTime and indicate how long it has been running for.

agrare commented on app/javascript/components/request-workflow-status/data.js:
```diff
@@ -59,6 +59,15 @@ export const workflowStatusData = (response) => {
   const rows = response.context ? rowData(response.context) : [];
+  if (response.context && response.context.State) {
+    rows.push({
+      id: `${response.context.State.Guid.toString()}+${Math.random()}`,
```
@GilbertCherrie why the Math.random() suffix?  Maybe related to another comment, we only want to append this if it doesn't exist in the StateHistory (if a workflow is complete, the last state will also be in StateHistory)

agrare replied:
Maybe when a workflow is complete we should null out the `context["State"]` ?  @Fryguy wdyt

Fryguy replied:
but regardless, the random number doesn't make sense to me.

GilbertCherrie replied:
Yeah that was just due to a duplicate state causing react warnings of duplicate ids

GilbertCherrie commented:
@Fryguy @agrare Here is a quick example:

<img width="1412" alt="Screenshot 2024-05-16 at 6 01 08 PM" src="https://github.com/ManageIQ/manageiq-ui-classic/assets/32444791/9c79980a-2618-4c27-b6fd-d87c1037e998">

Wanted to get your thoughts on this new row and icon choice

GilbertCherrie commented:
@Fryguy @agrare Here is a quick example:

<img width="1412" alt="Screenshot 2024-05-16 at 6 01 08 PM" src="https://github.com/ManageIQ/manageiq-ui-classic/assets/32444791/9c79980a-2618-4c27-b6fd-d87c1037e998">

Wanted to get your thoughts on this new row and the icon I chose.

agrare commented:
Looks good @GilbertCherrie I wonder if it's possible to format as `N days M hours T seconds` bit that's not critical

GilbertCherrie commented:
@agrare I made all the requested changes if you can review this again

agrare commented on app/javascript/components/request-workflow-status/data.js:
```diff
@@ -43,13 +43,39 @@ const convertDate = (date) => {
 
+// Converts the duration time in ms and returns a string in format: w days x hours y minutes z seconds
+// duration: time in ms
+const convertDuration = (duration) => {
```
Is there not a js library that converts to a "human readable" form?

Fryguy replied:
moment.js is the library.

GilbertCherrie replied:
I will try to rewrite it with moment.js

Fryguy replied:
`moment.duration(item.Duration).humanize()` is I think the entire thing.

Fryguy replied:
and it has locale support: `moment.duration(item.Duration).locale("en").humanize()`

Fryguy replied:
@GilbertCherrie Since some of the things run in sub-second, we might need to tweak the thresholds to humanize to not round to the minute. See https://momentjs.com/docs/#/durations/humanize/

Fryguy commented:
> Looks good @GilbertCherrie I wonder if it's possible to format as `N days M hours T seconds` bit that's not critical

More specifically the formatting should match whatever the other columns show.

Fryguy commented:
> @Fryguy @agrare Here is a quick example:
> <img alt="Screenshot 2024-05-16 at 6 01 08 PM" width="1412" src="https://private-user-images.githubusercontent.com/32444791/331397919-9c79980a-2618-4c27-b6fd-d87c1037e998.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU5Njc1MzYsIm5iZiI6MTcxNTk2NzIzNiwicGF0aCI6Ii8zMjQ0NDc5MS8zMzEzOTc5MTktOWM3OTk4MGEtMjYxOC00YzI3LWI2ZmQtZDg3YzEwMzdlOTk4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE3VDE3MzM1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRlYTM4ZjQ3MGFiNWY1NzMxMjQyOTA1MGQwZTQzZTY0NzBhNTU3ZjNiYTM5N2MyZDE0NmExOTljN2E3OWYzMWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OrThB8sovNCTvsx8OX851ltQG1PIG3gKIE3bEj3JFAg">
> 
> Wanted to get your thoughts on this new row and the icon I chose. For the duration I used Adam's suggestion of entered time - current time.

That's a good choice... I was also thinking something like possibly

- for running (with a yellow css color)
  - fa-spinner (specifically with a css rotation animation)
  - fa-ellipsis
  - fa-circle-play
- for completed (with a green css color - similar to the flash icon)
  - fa-circle-check 
- for failed (with a red css color - similar to the flash icon)
  - fa-circle-xmark

GilbertCherrie commented:
@agrare @Fryguy @jeffibm I fixed the icons if you can review again when you have time. Also, @Fryguy I changed the data table to use carbon icons so now this table uses carbon icons.

Fryguy commented on app/javascript/components/request-workflow-status/data.js:
```diff
@@ -43,13 +43,43 @@ const convertDate = (date) => {
+    resultString += `${durationString.slice(startIndex, durationString.indexOf('S') + 1)} `;
+    startIndex = durationString.indexOf('S') + 1;
+  }
+  return resultString;
```
As discussed, would it be possible to lowercase the string at the end?  I think `1h 45m 3s` looks better than `1H 45M 3S`, personally.

GilbertCherrie replied:
Yeah I can do that

Fryguy commented on app/javascript/components/request-workflow-status/data.js:
```diff
@@ -43,13 +43,43 @@ const convertDate = (date) => {
+// Converts the duration time in ms and returns a string in format: w days x hours y minutes z seconds
+// duration: time in ms
+const convertDuration = (duration) => {
+  const durationString = moment.duration(duration, 'milliseconds').toISOString().split('PT')[1];
```
I'm very surprised this doesn't "throw away" the days, months, years component of the time

Fryguy commented on app/javascript/components/request-workflow-status/data.js:
```diff
@@ -43,13 +43,43 @@ const convertDate = (date) => {
+};
+
+const getItemIcon = (item) => {
+  if (item.RunnerContext && item.RunnerContext.success) {
+    return { icon: 'carbon--CheckmarkOutline', props: { style: { color: '#2d7623' } } };
+  } if (item.RunnerContext && item.RunnerContext.Error) {
+    return { icon: 'carbon--MisuseOutline', props: { style: { color: '#cc0000' } } };
+  }
+  return { icon: 'carbon--PlayOutline', props: { style: { color: '#ec7a08' } } };
```
I think it might make more sense to use CSS here instead of inline styles?  I don't like the hardcoding of the colors if we have to change them later, it would be better in CSS.

GilbertCherrie replied:
Yeah I can change this, I saw the miq-data-table component takes in colour as a style prop so I thought we should use that but if css is preferred I can make the change

Fryguy commented on app/javascript/components/request-workflow-status/data.js:
```diff
@@ -59,6 +89,20 @@ export const workflowStatusData = (response) => {
+
+    rows.push({
+      id: state.Guid.toString(),
+      name: { text: state.Name, icon: 'carbon--PlayOutline', props: { style: { color: '#ec7a08' } } },
```
And it's also duplicated here.

Fryguy commented on app/javascript/components/request-workflow-status/data.js:
```diff
@@ -59,6 +89,20 @@ export const workflowStatusData = (response) => {
     return undefined;
   }
   const rows = response.context ? rowData(response.context) : [];
+  if (response.context && response.context.State) {
```
@agrare I think this code depends on nil-ing out the state when the flow is completed.  Is that in yet?

agrare replied:
No it isn't

Fryguy replied:
@GilbertCherrie Since the "final" state might appear twice, I'm concerned this will double up the presentation of it.  @agrare Thoughts on how we can avoid presenting the final state twice?

GilbertCherrie replied:
@Fryguy @agrare What if I add an ID check where if the current state id is found in an array of all the past ids it won't render the current state? The array might be a little large based on how many past states there is but its one possible solution if we want to handle the fix in the front end.

Fryguy replied:
@agrare If we do a state retry or a state gets re-run in a loop, does each iteration get it's own GUID?  If so, then yeah, an ID check might work.

@GilbertCherrie You could also just check the very last state in the StatesHistory and if it's the same as the "current" state then don't show it.

agrare replied:
> If we do a state retry or a state gets re-run in a loop, does each iteration get it's own GUID? If so, then yeah, an ID check might work.

Yes, every state execution gets a unique Guid even if it is retried.  Here is an example of the same state retried twice from floe specs:

```
"StateHistory"=>
    [{"Name"=>"State",
      "Input"=>{"foo"=>{"bar"=>"baz"}, "bar"=>{"baz"=>"foo"}},
      "Guid"=>"63837184-0795-41df-8d23-da034b8b2850",
      "EnteredTime"=>"2024-05-24T12:49:47Z",
      "RunnerContext"=>{"container_ref"=>"container-d"},
      "Output"=>{"Error"=>"States.Timeout"},
      "NextState"=>nil,
      "FinishedTime"=>"2024-05-24T12:49:47Z",
      "Duration"=>0.423689375,
      "RetryCount"=>2,
      "Retrier"=>["States.Timeout"]},
     {"Name"=>"State",
      "Input"=>{"foo"=>{"bar"=>"baz"}, "bar"=>{"baz"=>"foo"}},
      "Guid"=>"63837184-0795-41df-8d23-da034b8b2850",
      "EnteredTime"=>"2024-05-24T12:49:47Z",
      "RunnerContext"=>{"container_ref"=>"container-d"},
      "Output"=>{"Error"=>"States.Timeout"},
      "NextState"=>nil,
      "FinishedTime"=>"2024-05-24T12:49:47Z",
      "Duration"=>0.423689375,
      "RetryCount"=>2,
      "Retrier"=>["States.Timeout"]}]
```

Fryguy replied:
@agrare One more question.  When the entire flow is done, does the "current" state have a "finished time"?  If so, that might be the simplest check.

agrare replied:
Yes,
```
   "State"=>
    {"Name"=>"NextState",
     "Input"=>{"foo"=>1, "result"=>{"foo"=>"bar", "bar"=>"baz"}},
     "Guid"=>"81de8e05-8118-49d0-b97e-779904fb07c3",
     "EnteredTime"=>"2024-05-24T14:34:29Z",
     "Output"=>{"foo"=>1, "result"=>{"foo"=>"bar", "bar"=>"baz"}},
     "NextState"=>nil,
     "FinishedTime"=>"2024-05-24T14:34:29Z",
     "Duration"=>0.348082391}
```

Fryguy commented:
@GilbertCherrie I love the new icons, but there seems to be a spacing issue - I feel like the old icons looked proportionally better or perhaps aligned better?  I can't quite put my finger on it.

GilbertCherrie commented:
> @GilbertCherrie I love the new icons, but there seems to be a spacing issue - I feel like the old icons looked proportionally better or perhaps aligned better? I can't quite put my finger on it.

I can add padding to the right if they are too close

GilbertCherrie commented:
@Fryguy I made all the requested changes if you can review again when you have time

GilbertCherrie commented:
@Fryguy I made all the requested changes if you can review again when you have time

Fryguy commented:
Backported to `radjabov` in commit dc5281932c9300f0f49f55512a8a6d4b1b18ec33.

```text
commit dc5281932c9300f0f49f55512a8a6d4b1b18ec33
Author: Jeffrey bonson <jeffbonson@gmail.com>
Date:   Fri May 24 15:04:07 2024 +0530

    Merge pull request #9191 from GilbertCherrie/fix_workflow_states_table
    
    Add current state to workflow states table and format duration time
    
    (cherry picked from commit a9b704b98d530a01083ed8dc1cd3cf5eeeea5e47)
```
